// @mocko/proxy/dist/api/flag/flag.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlagService = void 0;
const service_1 = require("../../utils/decorators/service");
const flag_repository_1 = require("./flag.repository");
let FlagService = class FlagService {
    constructor(repository) {
        this.repository = repository;
    }
    setFlag(key, value) {
        this.repository.set(key, value);
    }
    getFlag(key) {
        return this.repository.get(key);
    }
    delFlag(key) {
        this.repository.del(key);
    }
    hasFlag(key) {
        return this.repository.has(key);
    }
};
FlagService = __decorate([
    service_1.Service(),
    __metadata("design:paramtypes", [flag_repository_1.FlagRepository])
], FlagService);
exports.FlagService = FlagService;
