// @mocko/proxy/dist/api/proxy/proxy.router.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProxyRouter = void 0;
const provider_1 = require("../../utils/decorators/provider");
const proxy_controller_1 = require("./proxy.controller");
let ProxyRouter = class ProxyRouter {
    constructor(controller) {
        this.controller = controller;
    }
    async getRoutes() {
        return [{
                method: '*',
                path: '/{any*}',
                handler: this.controller.proxyRequest.bind(this.controller),
                config: {
                    payload: {
                        output: 'stream',
                        parse: false
                    }
                }
            }];
    }
};
ProxyRouter = __decorate([
    provider_1.Provider(),
    __metadata("design:paramtypes", [proxy_controller_1.ProxyController])
], ProxyRouter);
exports.ProxyRouter = ProxyRouter;
